// @getlarge/nestjs-tools-file-storage/src/lib/file-storage.class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileStorage = void 0;
const defaultErrorMessage = 'Funtion must be implemented';
class FileStorage {
    constructor() {
        this.transformFilePath = (fileName, methodType, request, options = {}) => {
            throw new Error(defaultErrorMessage);
        };
    }
    fileExists(args) {
        throw new Error(defaultErrorMessage);
    }
    moveFile(args) {
        throw new Error(defaultErrorMessage);
    }
    uploadFile(args) {
        throw new Error(defaultErrorMessage);
    }
    uploadStream(args) {
        throw new Error(defaultErrorMessage);
    }
    downloadFile(args) {
        throw new Error(defaultErrorMessage);
    }
    downloadStream(args) {
        throw new Error(defaultErrorMessage);
    }
    deleteFile(args) {
        throw new Error(defaultErrorMessage);
    }
    getFileMeta(args) {
        throw new Error(defaultErrorMessage);
    }
    // TODO:
    // createDir(args: FileStorageDirBaseArgs): Promise<void> {
    //   throw new Error(defaultErrorMessage);
    // }
    deleteDir(args) {
        throw new Error(defaultErrorMessage);
    }
    readDir(args) {
        throw new Error(defaultErrorMessage);
    }
}
exports.FileStorage = FileStorage;
